src/colmap/sfm/incremental_mapper_impl.h

```diff
@@ -0,0 +1,100 @@
+
+namespace colmap {
+
+// Algorithm class for incremental mapper to make it easier to extend
+class IncrementalMapperImpl {
+ public:
+  // Find seed images for incremental reconstruction. Suitable seed images have
+  // a large number of correspondences and have camera calibration priors. The
+  // returned list is ordered such that most suitable images are in the front.
+  static std::vector<image_t> FindFirstInitialImage(
+      const IncrementalMapper::Options& options,
+      const std::shared_ptr<class Reconstruction>& reconstruction,
+      const std::shared_ptr<class ObservationManager>& obs_manager,
+      const std::unordered_map<image_t, size_t>& init_num_reg_trials,
+      const std::unordered_map<image_t, size_t>& num_registrations);
+
+  // For a given first seed image, find other images that are connected to the
+  // first image. Suitable second images have a large number of correspondences
+  // to the first image and have camera calibration priors. The returned list is
+  // ordered such that most suitable images are in the front.
+  static std::vector<image_t> FindSecondInitialImage(
+      const IncrementalMapper::Options& options,
+      image_t image_id1,
+      const std::shared_ptr<const DatabaseCache>& database_cache,
+      const std::shared_ptr<class Reconstruction>& reconstruction,
+      const std::unordered_map<image_t, size_t>& num_registrations);
+
+  // Implement IncrementalMapper::FindInitialImagePair
+  static bool FindInitialImagePair(
+      const IncrementalMapper::Options& options,
+      const std::shared_ptr<const DatabaseCache>& database_cache,
+      const std::shared_ptr<class Reconstruction>& reconstruction,
+      const std::shared_ptr<class ObservationManager>& obs_manager,
+      const std::unordered_map<image_t, size_t>& init_num_reg_trials,
+      const std::unordered_map<image_t, size_t>& num_registrations,
+      std::unordered_set<image_pair_t>& init_image_pairs,
+      TwoViewGeometry& two_view_geometry,
+      image_t& image_id1,
+      image_t& image_id2);
+
+  // Implement IncrementalMapper::FindNextImages
+  static std::vector<image_t> FindNextImages(
+      const IncrementalMapper::Options& options,
+      const std::shared_ptr<class Reconstruction>& reconstruction,
+      const std::shared_ptr<class ObservationManager>& obs_manager,
+      const std::unordered_set<image_t>& filtered_images,
+      std::unordered_map<image_t, size_t>& m_num_reg_trials);
+
+  // Implement IncrementalMapper::FindLocalBundle
+  static std::vector<image_t> FindLocalBundle(
+      const IncrementalMapper::Options& options,
+      image_t image_id,
+      const std::shared_ptr<class Reconstruction>& reconstruction);
+
+  // Implement IncrementalMapper::EstimateInitialTwoViewGeometry
+  static bool EstimateInitialTwoViewGeometry(
+      const IncrementalMapper::Options& options,
+      const std::shared_ptr<const DatabaseCache>& database_cache,
+      TwoViewGeometry& two_view_geometry,
```

Comment:
Suggest making the output argument always the last. In this case, we could also change the interface to return an std::optional<TwoViewGeometry> instead?

Reply:
Thanks. I have switched the order. Making it optional will also lead to changes in the code here: https://github.com/colmap/colmap/blob/main/src/colmap/controllers/incremental_pipeline.cc#L280-L300, where the two_view_geometry object is shared across different function calls. I would leave it as it is for this PR.